streams/src/main/java/org/apache/kafka/streams/processor/internals/CompositeRestoreListener.java

```diff
@@ -0,0 +1,118 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.processor.internals;
+
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.BatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.StateRestoreCallback;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+
+import java.util.Collection;
+
+class CompositeRestoreListener implements BatchingStateRestoreCallback, StateRestoreListener {
+
+    private final InternalStateRestoreAdapter stateRestoreAdapter;
+    private final StateRestoreListener storeRestoreListener;
+    private StateRestoreListener reportingStoreListener = new NoOpStateRestoreListener();
+
+    CompositeRestoreListener(StateRestoreCallback stateRestoreCallback) {
+
+        if (stateRestoreCallback instanceof StateRestoreListener) {
+            storeRestoreListener = (StateRestoreListener) stateRestoreCallback;
+        } else {
+            storeRestoreListener = new NoOpStateRestoreListener();
+        }
+
+        if (stateRestoreCallback == null) {
+            stateRestoreCallback = new NoOpStateRestoreCallback();
+        }
+
+        this.stateRestoreAdapter = new InternalStateRestoreAdapter(stateRestoreCallback);
+    }
+
+    @Override
+    public void onRestoreStart(TopicPartition topicPartition, String storeName,
+                               long startingOffset, long endingOffset) {
+        reportingStoreListener.onRestoreStart(topicPartition, storeName, startingOffset, endingOffset);
+        storeRestoreListener.onRestoreStart(topicPartition, storeName, startingOffset, endingOffset);
+    }
+
+    @Override
+    public void onBatchRestored(TopicPartition topicPartition, String storeName,
+                                long batchEndOffset, long numRestored) {
+        reportingStoreListener.onBatchRestored(topicPartition, storeName, batchEndOffset, numRestored);
+        storeRestoreListener.onBatchRestored(topicPartition, storeName, batchEndOffset, numRestored);
+    }
+
+    @Override
+    public void onRestoreEnd(TopicPartition topicPartition, String storeName,
+                             long totalRestored) {
+        reportingStoreListener.onRestoreEnd(topicPartition, storeName, totalRestored);
+        storeRestoreListener.onRestoreEnd(topicPartition, storeName, totalRestored);
+
+    }
+
+    @Override
+    public void restoreAll(final Collection<KeyValue<byte[], byte[]>> records) {
+        stateRestoreAdapter.restoreAll(records);
+    }
+
+    void setReportingStoreListener(StateRestoreListener reportingStoreListener) {
```

Comment:
Updated the set method to reflect the new name.  

Comments on the `composite` class and synchronization same as above.